utilities/ovs-appctl.8.in
.\" -*- nroff -*-
.so lib/ovs.tmac
.TH ovs\-appctl 8 "@VERSION@" "Open vSwitch" "Open vSwitch Manual"
.ds PN ovs\-appctl
.
.SH NAME
ovs\-appctl \- utility for configuring running Open vSwitch daemons
.
.SH SYNOPSIS
\fBovs\-appctl\fR [\fB\-\-target=\fItarget\fR | \fB\-t\fR \fItarget\fR]
[\fB\-T \fIsecs\fR | \fB\-\-timeout=\fIsecs\fR]
\fIcommand \fR[\fIarg\fR...]
.br
\fBovs\-appctl \fB\-\-help\fR
.br
\fBovs\-appctl \fB\-\-version\fR
.SH DESCRIPTION
Open vSwitch daemons accept certain commands at runtime to control their
behavior and query their settings.  Every daemon accepts a common set of
commands documented under \fBCOMMON COMMANDS\fR below.  Some daemons
support additional commands documented in their own manpages.
\fBovs\-vswitchd\fR in particular accepts a number of additional
commands documented in \fBovs\-vswitchd\fR(8).
.PP
The \fBovs\-appctl\fR program provides a simple way to invoke these
commands.  The command to be sent is specified on \fBovs\-appctl\fR's
command line as non-option arguments.  \fBovs\-appctl\fR sends the
command and prints the daemon's response on standard output.
.PP
In normal use only a single option is accepted:
.IP "\fB\-t \fItarget\fR"
.IQ "\fB\-\-target=\fItarget\fR"
Tells \fBovs\-appctl\fR which daemon to contact.
.IP
If \fItarget\fR begins with \fB/\fR it must name a Unix domain socket
on which an Open vSwitch daemon is listening for control channel
connections.  By default, each daemon listens on a Unix domain socket
named \fB@RUNDIR@/\fIprogram\fB.\fIpid\fB.ctl\fR, where \fIprogram\fR
is the program's name and \fIpid\fR is its process ID.  For example,
if \fBovs\-vswitchd\fR has PID 123, it would listen on
\fB@RUNDIR@/ovs\-vswitchd.123.ctl\fR.
.IP
Otherwise, \fBovs\-appctl\fR looks for a pidfile, that is, a file
whose contents are the process ID of a running process as a decimal
number, named \fB@RUNDIR@/\fItarget\fB.pid\fR.  (The \fB\-\-pidfile\fR
option makes an Open vSwitch daemon create a pidfile.)
\fBovs\-appctl\fR reads the pidfile, then looks for a Unix socket
named \fB@RUNDIR@/\fItarget\fB.\fIpid\fB.ctl\fR, where \fIpid\fR is
replaced by the process ID read from the pidfile, and uses that file
as if it had been specified directly as the target.
.IP
On Windows, \fItarget\fR can be an absolute path to a file that contains
a localhost TCP port on which an Open vSwitch daemon is listening
for control channel connections. By default, each daemon writes the
TCP port on which it is listening for control connection into the file
\fIprogram\fB.ctl\fR located inside the configured \fIOVS_RUNDIR\fR
directory. If \fItarget\fR is not an absolute path, \fBovs\-appctl\fR
looks for a file named \fItarget\fB.ctl\fR in the configured \fIOVS_RUNDIR\fR
directory.
.IP
The default target is \fBovs\-vswitchd\fR.
.
.IP "\fB\-T \fIsecs\fR"
.IQ "\fB\-\-timeout=\fIsecs\fR"
By default, or with a \fIsecs\fR of \fB0\fR, \fBovs\-appctl\fR waits
forever to connect to the daemon and receive a response.  This option
limits runtime to approximately \fIsecs\fR seconds.  If the timeout
expires, \fBovs\-appctl\fR exits with a \fBSIGALRM\fR signal.
.
.SH COMMON COMMANDS
Every Open vSwitch daemon supports a common set of commands, which are
documented in this section.
.
.SS GENERAL COMMANDS
These commands display daemon-specific commands and the running version.
Note that these commands are different from the \fB\-\-help\fR and
\fB\-\-version\fR options that return information about the
\fBovs\-appctl\fR utility itself.
.
.IP "\fBlist-commands\fR"
Lists the commands supported by the target.
.
.IP "\fBversion\fR"
Displays the version and compilation date of the target.
.
.SS LOGGING COMMANDS
Open vSwitch has several log levels.  The highest-severity log level is:
.
.IP "\fBoff\fR"
No message is ever logged at this level, so setting a logging
destination's log level to \fBoff\fR disables logging to that destination.
.
.PP
The following log levels, in order of descending severity, are
available:
.
.IP "\fBemer\fR"
A major failure forced a process to abort.
.IP "\fBerr\fR"
A high-level operation or a subsystem failed.  Attention is
warranted.
.IP "\fBwarn\fR"
A low-level operation failed, but higher-level subsystems may be able
to recover.
.IP "\fBinfo\fR"
Information that may be useful in retrospect when investigating
a problem.
.IP "\fBdbg\fR"
Information useful only to someone with intricate knowledge of the
system, or that would commonly cause too-voluminous log output.  Log
messages at this level are not logged by default.
.
.PP
Every Open vSwitch daemon supports the following commands for examining
and adjusting log levels.
.IP "\fBvlog/list\fR"
Lists the known logging modules and their current levels.
.
.IP "\fBvlog/list-pattern\fR"
Lists logging pattern used for each destination.
.
.IP "\fBvlog/set\fR [\fIspec\fR]"
Sets logging levels.  Without any \fIspec\fR, sets the log level for
every module and destination to \fBdbg\fR.  Otherwise, \fIspec\fR is a
list of words separated by spaces or commas or colons, up to one from
each category below:
.
.RS
.IP \(bu
A valid module name, as displayed by the \fBvlog/list\fR command on
\fBovs\-appctl\fR(8), limits the log level change to the specified
module.
.
.IP \(bu
\fBsyslog\fR, \fBconsole\fR, or \fBfile\fR, to limit the log level
change to only to the system log, to the console, or to a file,
respectively.
.IP
On Windows platform, \fBsyslog\fR is accepted as a word and
is only useful if the \fItarget\fR was started with the
\fB\-\-syslog\-target\fR option (the word has no effect otherwise).
.
.IP \(bu 
\fBoff\fR, \fBemer\fR, \fBerr\fR, \fBwarn\fR, \fBinfo\fR, or
\fBdbg\fR, to control the log level.  Messages of the given severity
or higher will be logged, and messages of lower severity will be
filtered out.  \fBoff\fR filters out all messages.
.RE
.
.IP
Case is not significant within \fIspec\fR.
.IP
Regardless of the log levels set for \fBfile\fR, logging to a file
will not take place unless the target application was invoked with the
\fB\-\-log\-file\fR option.
.IP
For compatibility with older versions of OVS, \fBany\fR is accepted as
a word but has no effect.
.
.IP "\fBvlog/set PATTERN:\fIdestination\fB:\fIpattern\fR"
Sets the log pattern for \fIdestination\fR to \fIpattern\fR.  Each time a
message is logged to \fIdestination\fR, \fIpattern\fR determines the
message's formatting.  Most characters in \fIpattern\fR are copied
literally to the log, but special escapes beginning with \fB%\fR are
expanded as follows:
.
.RS
.IP \fB%A\fR
The name of the application logging the message, e.g. \fBovs\-vswitchd\fR.
.
.IP \fB%B\fR
The RFC5424 syslog PRI of the message.
.
.IP \fB%c\fR
The name of the module (as shown by \fBovs\-appctl \-\-list\fR) logging
the message.
.
.IP \fB%d\fR
The current date and time in ISO 8601 format (YYYY\-MM\-DD HH:MM:SS).
.
.IP \fB%d{\fIformat\fB}\fR
The current date and time in the specified \fIformat\fR, which takes
the same format as the \fItemplate\fR argument to \fBstrftime\fR(3).
As an extension, any \fB#\fR characters in \fIformat\fR will be
replaced by fractional seconds, e.g. use \fB%H:%M:%S.###\fR for the
time to the nearest millisecond.  Sub-second times are only
approximate and currently decimal places after the third will always
be reported as zero.
.
.IP \fB%D\fR
The current UTC date and time in ISO 8601 format (YYYY\-MM\-DD HH:MM:SS).
.
.IP \fB%D{\fIformat\fB}\fR
The current UTC date and time in the specified \fIformat\fR, which
takes the same format as the \fItemplate\fR argument to
\fBstrftime\fR(3).  Supports the same extension for sub-second
resolution as \fB%d{\fR...\fB}\fR.
.
.IP \fB%E\fR
The hostname of the node running the application.
.
.IP \fB%m\fR
The message being logged.
.
.IP \fB%N\fR
A serial number for this message within this run of the program, as a
decimal number.  The first message a program logs has serial number 1,
the second one has serial number 2, and so on.
.
.IP \fB%n\fR
A new-line.
.
.IP \fB%p\fR
The level at which the message is logged, e.g. \fBDBG\fR.
.
.IP \fB%P\fR
The program's process ID (pid), as a decimal number.
.
.IP \fB%r\fR
The number of milliseconds elapsed from the start of the application
to the time the message was logged.
.
.IP \fB%t\fR
The subprogram name, that is, an identifying name for the process or
thread that emitted the log message, such as \fBmonitor\fR for the
process used for \fB\-\-monitor\fR or \fBmain\fR for the primary
process or thread in a program.
.
.IP \fB%T\fR
The subprogram name enclosed in parentheses, e.g. \fB(monitor)\fR, or
the empty string for the primary process or thread in a program.
.
.IP \fB%%\fR
A literal \fB%\fR.
.RE
.
.IP
A few options may appear between the \fB%\fR and the format specifier
character, in this order:
.
.RS
.IP \fB\-\fR
Left justify the escape's expansion within its field width.  Right
justification is the default.
.
.IP \fB0\fR
Pad the field to the field width with \fB0\fRs.  Padding with spaces
is the default.
.
.IP \fIwidth\fR
A number specifies the minimum field width.  If the escape expands to
fewer characters than \fIwidth\fR then it is padded to fill the field
width.  (A field wider than \fIwidth\fR is not truncated to fit.)
.RE
.
.IP
The default pattern for console and file output is \fB%D{%Y-%m-%dT
%H:%M:%SZ}|%05N|%c|%p|%m\fR; for syslog output, \fB%05N|%c|%p|%m\fR.
.
.IP
Daemons written in Python (e.g. \fBovs\-xapi\-sync\fR) do not allow
control over the log pattern.
.
.IP "\fBvlog/set\fR FACILITY:\fIfacility\fR"
Sets the RFC5424 facility of the log message. \fIfacility\fR can be one of
\fBkern\fR, \fBuser\fR, \fBmail\fR, \fBdaemon\fR, \fBauth\fR, \fBsyslog\fR,
\fBlpr\fR, \fBnews\fR, \fBuucp\fR, \fBclock\fR, \fBftp\fR, \fBntp\fR,
\fBaudit\fR, \fBalert\fR, \fBclock2\fR, \fBlocal0\fR, \fBlocal1\fR,
\fBlocal2\fR, \fBlocal3\fR, \fBlocal4\fR, \fBlocal5\fR, \fBlocal6\fR or
\fBlocal7\fR.
.
.IP "\fBvlog/close\fR"
Causes the daemon to close its log file, if it is open.  (Use
\fBvlog/reopen\fR to reopen it later.)
.
.IP "\fBvlog/reopen\fR"
Causes the daemon to close its log file, if it is open, and then
reopen it.  (This is useful after rotating log files, to cause a new
log file to be used.)
.IP
This has no effect if the target application was not invoked with the
\fB\-\-log\-file\fR option.
.
.SH OPTIONS
.
.so lib/common.man
.
.SH "SEE ALSO"
.
\fBovs\-appctl\fR can control all Open vSwitch daemons, including:
.BR ovs\-vswitchd (8),
and
.BR ovsdb\-server (8).


NEWS
v2.5.7 - xx xxx xxxx
--------------------

v2.5.6 - 19 Oct 2018
--------------------
   - Bug fixes
   - Bug fixes

v2.5.5 - 18 May 2018
---------------------
   - Bug fixes

v2.5.4 - 03 Sep 2017
---------------------
   - Bug fixes

v2.5.3 - 28 Jun 2017
---------------------
   - Bug fixes

v2.5.2 - 21 Feb 2017
---------------------
   - ovs-ctl:
     * Added the ability to selectively start the forwarding and database
       functions (ovs-vswitchd and ovsdb-server, respectively).
   - Fedora packaging:
     * A package upgrade does not automatically restart OVS service.
   - Bug fixes

v2.5.1 - 27 Sep 2016
---------------------
   - DPDK:
     * New appctl command 'dpif-netdev/pmd-rxq-show' to check the port/rxq
       assignment.
     * Type of log messages from PMD threads changed from INFO to DBG.
   - ovs-pki: Changed message digest algorithm from SHA-1 to SHA-512 because
     SHA-1 is no longer secure and some operating systems have started to
     disable it in OpenSSL.
   - Bug fixes

v2.5.0 - 26 Feb 2016
---------------------
   - Dropped support for Python older than version 2.7.  As a consequence,
     using Open vSwitch 2.5 or later on XenServer 6.5 or earlier (which
     have Python 2.4) requires first installing Python 2.7.
   - OpenFlow:
     * Group chaining (where one OpenFlow group triggers another) is
       now supported.
     * OpenFlow 1.4+ "importance" is now considered for flow eviction.
     * OpenFlow 1.4+ OFPTC_EVICTION is now implemented.
     * OpenFlow 1.4+ OFPTC_VACANCY_EVENTS is now implemented.
     * OpenFlow 1.4+ OFPMP_TABLE_DESC is now implemented.
     * Allow modifying the ICMPv4/ICMPv6 type and code fields.
     * OpenFlow 1.4+ OFPT_SET_ASYNC_CONFIG and OFPT_GET_ASYNC_CONFIG are
       now implemented.
   - ovs-ofctl:
     * New "out_group" keyword for OpenFlow 1.1+ matching on output group.
   - Tunnels:
     * Geneve tunnels can now match and set options and the OAM bit.
     * The nonstandard GRE64 tunnel extension has been dropped.
   - Support Multicast Listener Discovery (MLDv1 and MLDv2).
   - Add 'symmetric_l3l4' and 'symmetric_l3l4+udp' hash functions.
   - sFlow agent now reports tunnel and MPLS structures.
   - New 'check-system-userspace', 'check-kmod' and 'check-kernel' Makefile
     targets to run a new system testsuite.  These tests can be run inside
     a Vagrant box.  See INSTALL.md for details
   - Mark --syslog-target argument as deprecated.  It will be removed in
     the next OVS release.
   - Added --user option to all daemons
   - Add support for connection tracking through the new "ct" action
     and "ct_state"/"ct_zone"/"ct_mark"/"ct_label" match fields.  Only
     available on Linux kernels with the connection tracking module loaded.
   - Add experimental version of OVN.  OVN, the Open Virtual Network, is a
     system to support virtual network abstraction.  OVN complements the
     existing capabilities of OVS to add native support for virtual network
     abstractions, such as virtual L2 and L3 overlays and security groups.
   - RHEL packaging:
     * DPDK ports may now be created via network scripts (see README.RHEL).
   - DPDK:
     * Requires DPDK 2.2
     * Added multiqueue support to vhost-user
     * Note: QEMU 2.5+ required for multiqueue support
   - SELinux:
     * Introduced SELinux policy package.

v2.4.0 - 20 Aug 2015
---------------------
   - Flow table modifications are now atomic, meaning that each packet
     now sees a coherent version of the OpenFlow pipeline.  For
     example, if a controller removes all flows with a single OpenFlow
     "flow_mod", no packet sees an intermediate version of the OpenFlow
     pipeline where only some of the flows have been deleted.
   - Added support for SFQ, FQ_CoDel and CoDel qdiscs.
   - Add bash command-line completion support for ovs-vsctl Please check
     utilities/ovs-command-compgen.INSTALL.md for how to use.
   - The MAC learning feature now includes per-port fairness to mitigate
     MAC flooding attacks.
   - New support for a "conjunctive match" OpenFlow extension, which
     allows constructing OpenFlow matches of the form "field1 in
     {a,b,c...} AND field2 in {d,e,f...}" and generalizations.  For details,
     see documentation for the "conjunction" action in ovs-ofctl(8).
   - Add bash command-line completion support for ovs-appctl/ovs-dpctl/
     ovs-ofctl/ovsdb-tool commands.  Please check
     utilities/ovs-command-compgen.INSTALL.md for how to use.
   - The "learn" action supports a new flag "delete_learned" that causes
     the learned flows to be deleted when the flow with the "learn" action
     is deleted.
   - Basic support for the Geneve tunneling protocol. It is not yet
     possible to generate or match options. This is planned for a future
     release. The protocol is documented at
     http://tools.ietf.org/html/draft-gross-geneve-00
   - The OVS database now reports controller rate limiting statistics.
   - sflow now exports information about LACP-based bonds, port names, and
     OpenFlow port numbers, as well as datapath performance counters.
   - ovs-dpctl functionality is now available for datapaths integrated
     into ovs-vswitchd, via ovs-appctl.  Some existing ovs-appctl
     commands are now redundant and will be removed in a future
     release.  See ovs-vswitchd(8) for details.
   - OpenFlow:
     * OpenFlow 1.4 bundles are now supported for flow mods and port
       mods.  For flow mods, both 'atomic' and 'ordered' bundle flags
       are trivially supported, as all bundled messages are executed
       in the order they were added and all flow table modifications
       are now atomic to the datapath.  Port mods may not appear in
       atomic bundles, as port status modifications are not atomic.
     * IPv6 flow label and neighbor discovery fields are now modifiable.
     * OpenFlow 1.5 extended registers are now supported.
     * The OpenFlow 1.5 actset_output field is now supported.

utilities/ovs-appctl.c
/*
 * Copyright (c) 2008, 2009, 2010, 2011, 2012, 2014 Nicira, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at:
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <config.h>

#include <errno.h>
#include <getopt.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#include "command-line.h"
#include "daemon.h"
#include "dirs.h"
#include "openvswitch/dynamic-string.h"
#include "jsonrpc.h"
#include "process.h"
#include "timeval.h"
#include "unixctl.h"
#include "util.h"
#include "openvswitch/vlog.h"

static void usage(void);
static const char *parse_command_line(int argc, char *argv[]);
static struct jsonrpc *connect_to_target(const char *target);

int
main(int argc, char *argv[])
{
    char *cmd_result, *cmd_error;
    struct jsonrpc *client;
    char *cmd, **cmd_argv;
    const char *target;
    int cmd_argc;
    int error;

    set_program_name(argv[0]);

    /* Parse command line and connect to target. */
    target = parse_command_line(argc, argv);
    client = connect_to_target(target);

    /* Transact request and process reply. */
    cmd = argv[optind++];
    cmd_argc = argc - optind;
    cmd_argv = cmd_argc ? argv + optind : NULL;
    error = unixctl_client_transact(client, cmd, cmd_argc, cmd_argv,
                                    &cmd_result, &cmd_error);
    if (error) {
        ovs_fatal(error, "%s: transaction error", target);
    }

    if (cmd_error) {
        jsonrpc_close(client);
        fputs(cmd_error, stderr);
        ovs_error(0, "%s: server returned an error", target);
        exit(2);
    } else if (cmd_result) {
        fputs(cmd_result, stdout);
    } else {
        OVS_NOT_REACHED();
    }

    jsonrpc_close(client);
    free(cmd_result);
    free(cmd_error);
    return 0;
}

static void
usage(void)
{
    printf("\
%s, for querying and controlling Open vSwitch daemon\n\
usage: %s [TARGET] COMMAND [ARG...]\n\
Targets:\n\
  -t, --target=TARGET  pidfile or socket to contact\n\
Common commands:\n\
  list-commands      List commands supported by the target\n\
  version            Print version of the target\n\
  vlog/list          List current logging levels\n\
  vlog/list-pattern  List logging patterns for each destination.\n\
  vlog/set [SPEC]\n\
      Set log levels as detailed in SPEC, which may include:\n\
      A valid module name (all modules, by default)\n\
      'syslog', 'console', 'file' (all destinations, by default))\n\
      'off', 'emer', 'err', 'warn', 'info', or 'dbg' ('dbg', bydefault)\n\
  vlog/reopen        Make the program reopen its log file\n\
Other options:\n\
  --timeout=SECS     wait at most SECS seconds for a response\n\
  -h, --help         Print this helpful information\n\
  -V, --version      Display ovs-appctl version information\n",
           program_name, program_name);
    exit(EXIT_SUCCESS);
}

static const char *
parse_command_line(int argc, char *argv[])
{
    enum {
        OPT_START = UCHAR_MAX + 1,
        VLOG_OPTION_ENUMS
    };
    static const struct option long_options[] = {
        {"target", required_argument, NULL, 't'},
        {"execute", no_argument, NULL, 'e'},
        {"help", no_argument, NULL, 'h'},
        {"option", no_argument, NULL, 'o'},
        {"version", no_argument, NULL, 'V'},
        {"timeout", required_argument, NULL, 'T'},
        VLOG_LONG_OPTIONS,
        {NULL, 0, NULL, 0},
    };
    char *short_options_ = ovs_cmdl_long_options_to_short_options(long_options);
    char *short_options = xasprintf("+%s", short_options_);
    const char *target;
    int e_options;
    unsigned int timeout = 0;

    target = NULL;
    e_options = 0;
    for (;;) {
        int option;

        option = getopt_long(argc, argv, short_options, long_options, NULL);
        if (option == -1) {
            break;
        }
        switch (option) {
        case 't':
            if (target) {
                ovs_fatal(0, "-t or --target may be specified only once");
            }
            target = optarg;
            break;

        case 'e':
            /* We ignore -e for compatibility.  Older versions specified the
             * command as the argument to -e.  Since the current version takes
             * the command as non-option arguments and we say that -e has no
             * arguments, this just works in the common case. */
            if (e_options++) {
                ovs_fatal(0, "-e or --execute may be speciifed only once");
            }
            break;

        case 'h':
            usage();
            break;

        case 'o':
            ovs_cmdl_print_options(long_options);
            exit(EXIT_SUCCESS);

        case 'T':
            if (!str_to_uint(optarg, 10, &timeout) || !timeout) {
                ovs_fatal(0, "value %s on -T or --timeout is invalid", optarg);
            }
            break;

        case 'V':
            ovs_print_version(0, 0);
            exit(EXIT_SUCCESS);

        VLOG_OPTION_HANDLERS

        case '?':
            exit(EXIT_FAILURE);

        default:
            OVS_NOT_REACHED();
        }
    }
    free(short_options_);
    free(short_options);

    ctl_timeout_setup(timeout);

    if (optind >= argc) {
        ovs_fatal(0, "at least one non-option argument is required "
                  "(use --help for help)");
    }

    return target ? target : "ovs-vswitchd";
}

static struct jsonrpc *
connect_to_target(const char *target)
{
    struct jsonrpc *client;
    char *socket_name;
    int error;

#ifndef _WIN32
    if (target[0] != '/') {
        char *pidfile_name;
        pid_t pid;

        pidfile_name = xasprintf("%s/%s.pid", ovs_rundir(), target);
        pid = read_pidfile(pidfile_name);
        if (pid < 0) {
            ovs_fatal(-pid, "cannot read pidfile \"%s\"", pidfile_name);
        }
        free(pidfile_name);
        socket_name = xasprintf("%s/%s.%ld.ctl",
                                ovs_rundir(), target, (long int) pid);
#else
    /* On windows, if the 'target' contains ':', we make an assumption that
     * it is an absolute path. */
    if (!strchr(target, ':')) {
        socket_name = xasprintf("%s/%s.ctl", ovs_rundir(), target);
#endif
    } else {
        socket_name = xstrdup(target);
    }

    error = unixctl_client_create(socket_name, &client);
    if (error) {
        ovs_fatal(error, "cannot connect to \"%s\"", socket_name);
    }
    free(socket_name);

    return client;
}

